src/main/java/uk/co/ramp/covid/simulation/lockdown/LockdownController.java

```diff
@@ -7,105 +7,54 @@
         population = p;
+        components = new ArrayList<>();
     }
     
-    public void setLockdown(Time start, Time end, double sdist) {
-        lockdownStart = start;
-        lockdownEnd = end;
-        socialDist = sdist;
-    }
-
-    public void setSchoolLockdown(Time start, Time end, double sdist) {
-        lockdownStart = start;
-        lockdownEnd = end;
-        socialDist = sdist;
-        schoolLockdown = true;
-    }
-
-    public void implementLockdown(Time now) {
-        if (now.equals(lockdownStart)) {
-            enterLockdown(now);
-        }
-
-        if (now.equals(lockdownEnd)) {
-            exitLockdown(now);
-        }
+    public LockdownController addComponent(LockdownComponent c) {
+        components.add(c);
+        return this;
     }
     
-    private void furloughStaff() {
-        for (Person p : population.getAllPeople()) {
-            p.furlough();
-        }
-    }
-
-    private void unFurloughStaff() {
-        for (Person p : population.getAllPeople()) {
-            p.unFurlough();
-        }
+    public void setLockdown(Time start, Time end, double sdist) {
+        addComponent(new FullLockdownComponent(start, end, population, sdist));
     }
     
-    private void lockdownPlaces() {
-        for (CommunalPlace cPlace : population.getPlaces().getAllPlaces()) {
-            cPlace.enterLockdown(socialDist);
-        }
-    }
-
-    private void unLockdownPlaces() {
-        for (CommunalPlace cPlace : population.getPlaces().getAllPlaces()) {
-            cPlace.exitLockdown();
-        }
-    }
-
-    private void enterLockdown(Time t) {
-        inLockdown = true;
-        lockdownPlaces();
-        furloughStaff();
-    }
-
-    // This sets the schools exempt from lockdown if that is triggered. Somewhat fudged at present by setting the schools to be KeyPremises
-    private void schoolExemption() {
-        for (School s : population.getPlaces().getSchools()) {
-            s.overrideKeyPremises(true);
-        }
-        for (Nursery n : population.getPlaces().getNurseries()) {
-            n.overrideKeyPremises(true);
-        }
+    public void setSchoolLockdown(Time start, Time end, Time schoolsReturn, double sdist) {
+        addComponent(new FullLockdownComponent(start, end, population, sdist));
+        addComponent(new SchoolReturnComponent(schoolsReturn, end, population));
     }
 
-    private void exitLockdown(Time t) {
-        // TODO: School lockdown still needs fixed this is the original "population" behaviour
-        if (schoolLockdown) {
-            schoolExemption();
-        } else {
-            unLockdownPlaces();
-            unFurloughStaff();
-            inLockdown = false;
+    public void implementLockdown(Time now) {
+        List<LockdownComponent> removeable = new ArrayList<>();
+        for (LockdownComponent c : components) {
+            c.handleLockdown(now);
+            
+            if (c.getEndTime().equals(now)) {
```

Comment:
The lockdown component is checking its own end time, so I'm not sure why we have something like this outside of the components.

Reply:
This is just a performance optimisation that let's us remove components that never need to be called again (and even at that it probably doesn't save us a huge amount!)